Add timer model, relationships- belongs_to :player, has_many :timers

PamBWillenz opened the pull request:
Next will do start, stop methods with a reset in the timer model. Then will write tests. I pushed code up - tests are not complete and not passing. This is a WIP. I will try to fix if I get feedback in next hour or two. I will be gone for weekend and will hit this when I return

PamBWillenz commented:
Adjusted tests but couldn't run rspec - tried bundle install, bundle update because first errors had to do with gem dependencies. Here is the error I currently have:
``` /vagrant/src/chess $ rspec spec models/timer_spec.rb
/home/vagrant/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/activesupport-4.1.9/lib/active_support/dependencies.rb:241:in `load': cannot load such file -- /vagrant/src/chess/models/timer_spec.rb (LoadError)
	from /home/vagrant/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/activesupport-4.1.9/lib/active_support/dependencies.rb:241:in `block in load'
	from /home/vagrant/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/activesupport-4.1.9/lib/active_support/dependencies.rb:232:in `load_dependency'
	from /home/vagrant/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/activesupport-4.1.9/lib/active_support/dependencies.rb:241:in `load'
	from /home/vagrant/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/configuration.rb:1435:in `block in load_spec_files'
	from /home/vagrant/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/configuration.rb:1433:in `each'
	from /home/vagrant/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/configuration.rb:1433:in `load_spec_files'
	from /home/vagrant/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:100:in `setup'
	from /home/vagrant/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:86:in `run'
	from /home/vagrant/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:71:in `run'
	from /home/vagrant/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/lib/rspec/core/runner.rb:45:in `invoke'
	from /home/vagrant/.rbenv/versions/2.3.0/lib/ruby/gems/2.3.0/gems/rspec-core-3.5.4/exe/rspec:4:in `<top (required)>'
	from /home/vagrant/.rbenv/versions/2.3.0/bin/rspec:23:in `load'
	from /home/vagrant/.rbenv/versions/2.3.0/bin/rspec:23:in `<main>'
```
Will try again tomorrow :(

FsDevNinja commented:
You need to clean up the schema file. if shouldn't have any <<<<<<< or ======== left after the merge conflict.

PamBWillenz commented:
Yep - ok - I didn't check the whole schema file - crap on a cracker.

PamBWillenz commented:
Having an issue with running rspec - same error as above (previous comment) appreciate any help tomorrow. I can't watch CI fail again.

FsDevNinja commented:
Since we reverted the schema file we need to fix our test database.  rake db:migrate RAILS_ENV=test
This should fix your issue. You got it Pam!!! I had to fix the same stuff earlier.

lasley commented on app/models/timer.rb:
```diff
@@ -0,0 +1,48 @@
+class Timer < ActiveRecord::Base
+
+  belongs_to :player
+
```
Needs to belong to game too?

lasley commented on app/models/timer.rb:
```diff
@@ -0,0 +1,48 @@
+
+
+  def start_time
+    Timer.create(player_id: current_player.id, time: Time.now)
```
This is creating a new timer each start, is that the intent?

This could be useful from a historical logging perspective, but the reset method kind of conflicts with that intent. If you reset, then start, all the timers in history are zero.

PamBWillenz replied:
Yes, need to rework this because that isn't the intent (I think) - even though each new game will use new timers, but not each start of the timer because a player in a game can hit the start/stop several times during a game.

lasley commented on app/models/timer.rb:
```diff
@@ -0,0 +1,48 @@
+  end
+
+  def stop_time
+    last_attempt = Timer.where(player_id: current_player.id).last_attempt
```
Where is `Timer.last_attempt` defined?

PamBWillenz replied:
Yep - good point - need to rework.

lasley commented on app/models/timer.rb:
```diff
@@ -0,0 +1,48 @@
+    @total_time = 0
+    return r
+  end
+end
```
https://robots.thoughtbot.com/no-newline-at-end-of-file

PamBWillenz replied:
Thinking that a reset method (by itself) is unnecessary. I'll clean this up.

PamBWillenz commented:
I am trying many things to get the timer model and rspec tests working.

lasley commented on app/models/timer.rb:
```diff
@@ -0,0 +1,48 @@
+    @stop_time = nil
+    @running = false
+    @time_left = 0
+    @minutes = seconds / 60
```
Where is `seconds` defined?

lasley commented on app/models/timer.rb:
```diff
@@ -0,0 +1,48 @@
+    @running = false
+    @time_left = 0
+    @minutes = seconds / 60
+    @seconds = minutes * 60
```
Why `minutes * 60` after already `seconds / 60`?

PamBWillenz commented:
Add columns to timers table to store seconds/minutes in time_left versus using just time so the model works more like a countdown timer (vs a clock). The tests are passing. I added a test condition that could reflect the initial amount of time the players set for themselves as their countdown time - this test didn't pass. It kept getting nil - maybe have to initiate a number to set. I refactored some code to get tests to pass - took out `@time_left = self-time.left` and `save`.

PamBWillenz commented on spec/models/timer_spec.rb:
```diff
@@ -0,0 +1,30 @@
+    end
+
+    it "should determine how much time the players have" do
+      # expect(@time_left).to eq(Integer)
```
That could work, Kelly - so something like this:
```let(:player) { FactoryGirl.create(:player, current_player.id, email: 'pam@rook.com', password: 'BESTKITTY') }
    let(:game) { FactoryGirl.create(:game, game.id)}
   let(:timer) { FactoryGirl.create(:timer) }```

kellylougheed replied:
Yeah, that's what I was thinking!

PamBWillenz replied:
Created the above line - `let(:timer) { FactoryGirl.create(:timer) }` and keep getting this error: 
```
Failures:

  1) Timer should start timer for current players
     Failure/Error: let(:timer) { FactoryGirl.create(:timer)}
     
     ActiveRecord::RecordInvalid:
       Validation failed: Email can't be blank, Password can't be blank
     # ./spec/models/timer_spec.rb:8:in `block (2 levels) in <top (required)>'
     # ./spec/models/timer_spec.rb:20:in `block (2 levels) in <top (required)>'

  2) Timer should stop timer for current players
     Failure/Error: let(:timer) { FactoryGirl.create(:timer)}
     
     ActiveRecord::RecordInvalid:
       Validation failed: Email can't be blank, Password can't be blank
     # ./spec/models/timer_spec.rb:8:in `block (2 levels) in <top (required)>'
     # ./spec/models/timer_spec.rb:25:in `block (2 levels) in <top (required)>'

Finished in 0.83915 seconds (files took 3.77 seconds to load)
34 examples, 2 failures
```
I commented out creating a new timer for each test. I also put email and password in the FactoryGirl.create(:timer, email: 'pam@rook.com', password: 'BIGKITTY' to see if that would work. It didn't.

PamBWillenz replied:
Thanks, Kelly - I tried all variations of :player, current_player.id. I didn't try the one you put :). I'll try that. Yes, player_id should be the right attribute since a timer belongs_to a player.

PamBWillenz replied:
Kelly - this worked for these failing tests (yea!) But I still have 2 failing tests because I added an after_initialize callback - ugh.

lasley commented on app/models/timer.rb:
```diff
@@ -0,0 +1,31 @@
+  belongs_to :player
+  belongs_to :game
+
+  def initialize
```
This is breaking the inheritance chain for the ActiveRecord superclass. You either need to call the super, or (better) use a `after_initialize` callback instead.

PamBWillenz replied:
Ok - cool. I added these two - not sure which one is right - 
```
after_initialize :set_time, :if => :new_record?

  def set_time
    self.time ||= :timer
  end

  # after_initialize do |timer|
  # end
```

PamBWillenz replied:
After researching this more, I think I wrote this `after_initialize callback` more accurately, yet it is causing 2 failing tests :(. 

```
1) Timer should start timer for current players
     Failure/Error: self.timer ||= Timer.new
     
     NoMethodError:
       undefined method `timer' for #<Timer:0x007f564f3edbe8>
       Did you mean?  time
                      time=
                      time?
     # ./app/models/timer.rb:9:in `add_timer'
     # ./spec/models/timer_spec.rb:7:in `block (2 levels) in <top (required)>'
     # ./spec/models/timer_spec.rb:24:in `block (2 levels) in <top (required)>'

  2) Timer should stop timer for current players
     Failure/Error: self.timer ||= Timer.new
     
     NoMethodError:
       undefined method `timer' for #<Timer:0x007f564d663fa0>
       Did you mean?  time
                      time=
                      time?
     # ./app/models/timer.rb:9:in `add_timer'
     # ./spec/models/timer_spec.rb:7:in `block (2 levels) in <top (required)>'
     # ./spec/models/timer_spec.rb:29:in `block (2 levels) in <top (required)>'

Finished in 0.88573 seconds (files took 3.86 seconds to load)
34 examples, 2 failures

Failed examples:

rspec ./spec/models/timer_spec.rb:22 # Timer should start timer for current players
rspec ./spec/models/timer_spec.rb:27 # Timer should stop timer for current players

```

Here is the `after_initialize callback` I used - 

```
after_initialize :add_timer, :if => :new_record?

  def add_timer
    self.timer ||= Timer.new 
  end
```

lasley commented on app/models/timer.rb:
```diff
@@ -0,0 +1,31 @@
+  end
+
+  def start
+    @time_left 
```
What is the purpose of this line?

PamBWillenz replied:
This line sets how much time left on the timer as you go through your turns. If you start with 5 minutes, take a turn, and now have 3 minutes, @time_left should keep track of how much time you have left from each turn during that game. At least that is what I hope it does.

PamBWillenz replied:
I made changes but won't push yet since I'd like to solve these current failing tests.

PamBWillenz replied:
Thanks - I remember setting this and have tried again with the new associations - 
```
def start
    @time_left = self.time_left
    @running = true 
    @start_time = Time.now.round(0)
  end
```

PamBWillenz replied:
I am trying to get the tests to pass before I push up to the repo.

lasley commented on spec/models/timer_spec.rb:
```diff
@@ -0,0 +1,30 @@
+      timer.start
+      expect(timer.stop).to eq(Time.now.round(0))
+    end
+
```
Please add a test case for multiple start/stops

PamBWillenz replied:
Yes, I will after working out other errors.

PamBWillenz commented:
After trying many variations of code, I got the tests to pass. I also checked in the console that `timer` indeed was getting initiated (that was one of the issues I couldn't get working). I had to put a number in the setting the amount of time players wanted as their set time because I kept getting nil. Not sure if it is ok to hard code a number here or if there is a way to make this dynamic - Phew - this was a challenge :)

FsDevNinja commented:
@PamBWillenz 2 things i see. **1** I we need to get rid of this coverage stuff from making its way into master. run `git rm coverage -r` and commit results. **2** I see that you have a migration file missing from this branch. I see one for adding columns to the timer table but don't see one that is creating the table.

lasley commented:
I would guess that @FsDevNinja num 1 point will be solved after merging master in & resolving conflict.

lasley commented on coverage/.last_run.json:
```diff
@@ -0,0 +1,5 @@
+{
```
Coverage files made it back in during the conflict resolutions and need to be removed

lasley commented on app/models/timer.rb:
```diff
@@ -0,0 +1,26 @@
+class Timer < ActiveRecord::Base
+  after_initialize do |timer|
```
What is the purpose of this empty block?

lasley commented on app/models/timer.rb:
```diff
@@ -0,0 +1,26 @@
+  def start
+    @time_left = self.time_left
+    @running = true 
+    @start_time = Time.now.round(0)
```
Why rounding?

lasley commented on coverage/.resultset.json:
```diff
@@ -0,0 +1,419 @@
+{
```
Remove

lasley commented on spec/models/timer_spec.rb:
```diff
@@ -0,0 +1,40 @@
+  let(:timer) { FactoryGirl.create(:timer, player_id: player.id)}
+
+    it "should determine when a player starts" do 
+      expect(DateTime.new).to be_a(DateTime)
```
This is testing the `DateTime` module & is unnecessary.

lasley commented on spec/models/timer_spec.rb:
```diff
@@ -0,0 +1,40 @@
+
+    it "should determine how much time the players have" do
+      @time_left = 500
+      expect(@time_left).to eq(500)
```
This is a false positive. You're setting the `@time_left` to 500 right above this, there is no way for it to not be equal to 500

Also, you would want to be checking `timer.time_left`

lasley commented on spec/models/timer_spec.rb:
```diff
@@ -0,0 +1,40 @@
+    end
+    
+    it "should start timer for current players" do
+      expect(timer.start).to eq(Time.now.round(0))
```
Wouldn't you want to call `timer.start`, then check `timer.start_time`?

lasley commented on spec/models/timer_spec.rb:
```diff
@@ -0,0 +1,40 @@
+
+    it "should stop timer for current players" do 
+      timer.start
+      expect(timer.stop).to eq(Time.now.round(0))
```
It would be better to check the instance variable and not the return of `stop`. Although really both should be tested, assuming that you plan on using the return from stop (which seems a little unnecessary IMO)

lasley commented on app/models/timer.rb:
```diff
@@ -0,0 +1,26 @@
+  belongs_to :player
+  belongs_to :game
+
+  def start
```
`start!`

lasley commented on app/models/timer.rb:
```diff
@@ -0,0 +1,26 @@
+    @start_time = Time.now.round(0)
+  end
+  
+  def stop
```
`stop!`

lasley commented on spec/models/timer_spec.rb:
```diff
@@ -0,0 +1,40 @@
+      expect(timer.start).to eq(Time.now.round(0))
+      timer.stop
+      expect(timer.start).to eq(Time.now.round(0))
+      timer.stop
```
The stop is the more important thing to be testing here, which has the math differential as time goes on

PamBWillenz commented:
@lasley Dave - Took out after_initialize block, added ! to start and stop methods, worked through each of your test questions - now all but one test is failing. Most due to getting `nil` instead of a value or true/false. A couple of developers said the time cop gem is needed to really test time in motion. I worked through the tests but am stuck getting them to pass. I'd appreciate any advice/direction to get these to work. 

Here are the errors I am getting in my terminal:
``` Failures:

  1) Timer should determine how much time the players have
     Failure/Error: expect(timer.time_left).to eq(500)
     
       expected: 500
            got: nil
     
       (compared using ==)
     # ./spec/models/timer_spec.rb:13:in `block (2 levels) in <top (required)>'

  2) Timer should start timer for current players
     Failure/Error: expect(timer.start_time).to eq(now)
     
       expected: 2016-11-30 19:20:10.689625219 +0000
            got: nil
     
       (compared using ==)
     # ./spec/models/timer_spec.rb:19:in `block (2 levels) in <top (required)>'

  3) Timer should stop timer for current players
     Failure/Error: expect(@running).to eq(false)
     
       expected: false
            got: nil
     
       (compared using ==)
     # ./spec/models/timer_spec.rb:25:in `block (2 levels) in <top (required)>'

  4) Timer should have multiple starts and stops for players during a game
     Failure/Error: expect(@running).to eq(true) #can only test for this because can't time travel
     
       expected: true
            got: nil
     
       (compared using ==)
     # ./spec/models/timer_spec.rb:30:in `block (2 levels) in <top (required)>'

Finished in 3.14 seconds (files took 4.07 seconds to load)
63 examples, 4 failures

Failed examples:

rspec ./spec/models/timer_spec.rb:12 # Timer should determine how much time the players have
rspec ./spec/models/timer_spec.rb:16 # Timer should start timer for current players
rspec ./spec/models/timer_spec.rb:22 # Timer should stop timer for current players
rspec ./spec/models/timer_spec.rb:28 # Timer should have multiple starts and stops for players during a game

```

lasley commented on spec/models/timer_spec.rb:
```diff
@@ -0,0 +1,42 @@
+    it "should stop timer for current players" do 
+      timer.start!
+      timer.stop!
+      expect(@running).to eq(false)
```
Where is `@running` declared?

lasley commented on spec/models/timer_spec.rb:
```diff
@@ -0,0 +1,42 @@
+
+    it "should have multiple starts and stops for players during a game" do 
+      timer.start!
+      expect(@running).to eq(true) 
```
Where is `@running` declared?

lasley commented on spec/models/timer_spec.rb:
```diff
@@ -0,0 +1,42 @@
+    end
+
+    it "should end game for a player" do 
+      expect(@time_left).to eq nil
```
Where is `@time_left ` declared?

lasley commented on app/models/timer.rb:
```diff
@@ -0,0 +1,23 @@
+  belongs_to :game
+
+  def start!
+    @time_left = self.time_left
```
This just makes the variable equal to itself & can likely be removed. `@` is basically a shortcut for `self`

lasley commented on app/models/timer.rb:
```diff
@@ -0,0 +1,23 @@
+    if @running
+      @stop_time = Time.now
+      @running = false
+      self.time_left = @time_left.to_i - elapsed_time.to_i
```
`@time_left = `

PamBWillenz commented:
All the tests are passing - Matt helped me troubleshoot failing tests and refactoring. Huge help! Thanks, Matt! Not sure why coveralls is giving me an 'x'.

FsDevNinja commented:
Looks Good. once @lasley has his changes satisfied.

lasley commented:
Dismissed my review and merged
